### PR TITLE
Exclude files that do not satisfy the metadata predicate

### PR DESCRIPTION
**Pull request by chenjian2664**
## Description

Exclude files that do not satisfy the metadata predicate while computing the table statistics in Iceberg.
While building the tableScan metadata predicates are ignored, it results scan files more than expectation.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 1]
```diff
@@ -762,4 +764,15 @@ public static Map<String, Integer> columnNameToPositionInSchema(Schema schema)
                 (column, position) -> immutableEntry(column.name(), Long.valueOf(position).intValue()))
                 .collect(toImmutableMap(Entry::getKey, Entry::getValue));
     }
+
+    public static Domain getPathDomain(TupleDomain<IcebergColumnHandle> effectivePredicate)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:768]: Please add a preparatory commit to move this method.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java [hunk 2]
```diff
@@ -140,14 +144,22 @@ public static TableStatistics makeTableStatistics(
...
+        effectivePredicate = effectivePredicate.filter((icebergColumnHandle, domain) -> !icebergColumnHandle.equals(pathColumnHandle()));
         TableScan tableScan = icebergTable.newScan()
                 .filter(toIcebergExpression(effectivePredicate))
                 .useSnapshot(snapshotId)
                 .includeColumnStats();
 
         IcebergStatistics.Builder icebergStatisticsBuilder = new IcebergStatistics.Builder(columns, typeManager);
         try (CloseableIterable<FileScanTask> fileScanTasks = tableScan.planFiles()) {
-            fileScanTasks.forEach(fileScanTask -> icebergStatisticsBuilder.acceptDataFile(fileScanTask.file(), fileScanTask.spec()));
+            fileScanTasks.forEach(fileScanTask -> {
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java:156]: Could you add a test?

Comment (chenjian2664) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java:156]: Could you suggest where should the test be added? And do you have suggestion about how the test would looks like?

**Comment by findinpath** [comment]
```
trino> create table iceberg.default.ptest1(c1 integer, c2 integer, p integer) with (partitioning = array['p'], format='parquet');
CREATE TABLE
trino> insert into iceberg.default.ptest1 values (4,5, 10), (11,16,20);
INSERT: 2 rows

trino> insert into iceberg.default.ptest1 values (1,2, 10), (2,3,10), (12,14,20);


rino> show stats for (select * from iceberg.default.ptest1 where "$path" = 'hdfs://hadoop-master:9000/user/hive/warehouse/ptest1-5f9f47e5f84b4c66a25017e3952ecb0d/data/p=10/20230703_114010_00002_wmerg-de99aaa0-fa75-441a-ad87-174e6a86e50a.parquet');
 column_name | data_size | distinct_values_count | nulls_fraction | row_count | low_value | high_value 
-------------+-----------+-----------------------+----------------+-----------+-----------+------------
 c1          |      NULL |                   2.0 |            0.0 |      NULL | 1         | 2          
 c2          |      NULL |                   2.0 |            0.0 |      NULL | 2         | 3          
 p           |      NULL |                   1.0 |            0.0 |      NULL | 10        | 10         
 NULL        |      NULL |                  NULL |           NULL |       2.0 | NULL      | NULL       
(4 rows)

Query 20230703_121334_00025_wmerg, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230703_121334_00025_wmerg
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 0% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak Memory: 546B
25.82 [0 rows, 0B] [0 rows/s, 0B/s]

trino> show stats for (select * from iceberg.default.ptest1 where "$path" LIKE 'hdfs://hadoop-master:9000/user/hive/warehouse/ptest1-5f9f47e5f84b4c66a25017e3952ecb0d/data/p=10/%');
 column_name | data_size | distinct_values_count | nulls_fraction | row_count | low_value | high_value 
-------------+-----------+-----------------------+----------------+-----------+-----------+------------
 c1          |      NULL |                  NULL |           NULL |      NULL | NULL      | NULL       
 c2          |      NULL |                  NULL |           NULL |      NULL | NULL      | NULL       
 p           |      NULL |                  NULL |           NULL |      NULL | NULL      | NULL       
 NULL        |      NULL |                  NULL |           NULL |      NULL | NULL      | NULL       
(4 rows)

```


Retrieving stats for a single file works.
Retrieving stats for the entire partition doesn't seem to work.

Iceberg connector provides the stats as expected through the changes of @chenjian2664 .
However `io.trino.cost.FilterStatsCalculator.FilterExpressionStatsCalculatingVisitor#process` dismisses them.

@raunaqmorarka  do you happen to know why this is happening?

**Comment by findepi** [comment]
> Iceberg connector provides the stats as expected through the changes of @chenjian2664 . However `io.trino.cost.FilterStatsCalculator.FilterExpressionStatsCalculatingVisitor#process` dismisses them.
> 
> do you happen to know why this is happening?


Iceberg doesn't produce stats for `$path` column, co the engine cannot determine stats for a predicate involving the `$path`. The way to solve this would be to subsume the `$path` predicate into Iceberg connector, which can sometimes be done.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java [hunk 3]
```diff
@@ -140,14 +144,22 @@ public static TableStatistics makeTableStatistics(
                 .map(column -> Maps.immutableEntry(column.fieldId(), column.type()))
                 .collect(toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
 
+        Domain pathDomain = getPathDomain(effectivePredicate);
+        effectivePredicate = effectivePredicate.filter((icebergColumnHandle, domain) -> !icebergColumnHandle.equals(pathColumnHandle()));
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java:148]: this should use `.filter((column, domain) -> !isMetadataColumnId(column.getId()));`

https://github.com/trinodb/trino/blob/037118255a08b45b7a061991cd22628ff3ce01d4/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java#L157

add a test for stats with a predicate on the `$file_modified_time` column

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java [hunk 4]
```diff
@@ -140,6 +151,8 @@ public static TableStatistics makeTableStatistics(
                 .map(column -> Maps.immutableEntry(column.fieldId(), column.type()))
                 .collect(toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
 
+        Domain pathDomain = getPathDomain(effectivePredicate);
+        Domain fileModifiedTimeDomain = getFileModifiedTimePathDomain(effectivePredicate);
         TableScan tableScan = icebergTable.newScan()
                 // Table enforced constraint may include eg $path column predicate which is not handled by Iceberg library TODO apply $path and $file_modified_time filters here
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java:157]: This TODO can be removed now right?

cc @findepi

Comment (chenjian2664) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java]: Removed

**Comment by mosabua** [comment]
@findinpath and @chenjian2664 could you connect and figure out next steps here?

**Comment by findinpath** [comment]
@chenjian2664  pls rebase and I'll make another round of review. Appologies for not following up with the review on the last request.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java [hunk 5]
```diff
@@ -140,15 +151,28 @@ public static TableStatistics makeTableStatistics(
...
-                // Table enforced constraint may include eg $path column predicate which is not handled by Iceberg library TODO apply $path and $file_modified_time filters here
                 .filter(toIcebergExpression(effectivePredicate.filter((column, domain) -> !isMetadataColumnId(column.getId()))))
                 .useSnapshot(snapshotId)
                 .includeColumnStats();
 
         IcebergStatistics.Builder icebergStatisticsBuilder = new IcebergStatistics.Builder(columns, typeManager);
         try (CloseableIterable<FileScanTask> fileScanTasks = tableScan.planFiles()) {
-            fileScanTasks.forEach(fileScanTask -> icebergStatisticsBuilder.acceptDataFile(fileScanTask.file(), fileScanTask.spec()));
+            fileScanTasks.forEach(fileScanTask -> {
+                if (!pathDomain.includesNullableValue(utf8Slice(fileScanTask.file().path().toString()))) {
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java:164]: optional: We could also add another test in `TestIcebergFileOperations` which would showcase that calling `SHOW STATS` with a filter on `$path` or `modified_time` domain shrinks very much the amount of data files to be taken into account for processing.

Comment (chenjian2664) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java:164]: @findinpath @mosabua Thanks for the attention

Comment (chenjian2664) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java]: I tried write the tests for `path` and `file_modified_time` column with `SHOW STATS`, but found that the files access almost the same. for the `file_modified_time` there are even more access with `INPUT_FILE_LAST_MODIFIED` type file operation.

**Comment by chenjian2664** [comment]
@ebyhr @findinpath Need your help to review on this pr

**Comment by mosabua** [comment]
I think this is still important to pursue. Hopefully @chenjian2664 can rebase and then @findinpath and others can follow up.

**Comment by chenjian2664** [comment]
Thanks @mosabua 
I believe it is ready to be reviewed @findinpath @ebyhr

**Comment by findepi** [comment]
from PR description:

> Fix #17451

i am not sure this is true. I failed to reproduce #17451 on current master.

**Comment by chenjian2664** [comment]
> from PR description:
> 
> > Fix #17451
> 
> i am not sure this is true. I failed to reproduce #17451 on current master.
> 
> Can you please update the PR title and description to reflect current state?

Yes, just notice the issue already has been fixed. 👍🏻 
Updated the description, appreciate it if could take time to look at it. thanks

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 6]
```diff
@@ -5516,6 +5516,208 @@ public void testOptimizeWithPathColumn()
...
+        // Make sure the whole table has stats
+        MaterializedResult tableStatistics = computeActual(collectingStatisticsSession, "SHOW STATS FOR (SELECT * FROM %s WHERE \"$path\" IS NOT NULL)".formatted(tableName));
+        MaterializedResult expectedTableStatistics =
+                resultBuilder(collectingStatisticsSession, VARCHAR, DOUBLE, DOUBLE, DOUBLE, DOUBLE, VARCHAR, VARCHAR)
+                        .row("id", null, 4.0, 0.0, null, "1", "4")
+                        .row("value", null, 3.0, 0.25, null, "1", "4")
+                        .row(null, null, null, null, 4.0, null, null)
+                        .build();
+        if (format == AVRO) {
+            expectedTableStatistics =
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:5544]: pls add a comment why we have different stats when working with avro

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 7]
```diff
@@ -2801,7 +2801,7 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
                         Optional.empty()), // forAnalyze does not affect stats
                 handle -> {
                     Table icebergTable = catalog.loadTable(session, handle.getSchemaTableName());
-                    return TableStatisticsReader.getTableStatistics(typeManager, session, handle, icebergTable);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:2804]: @chenjian2664 `Fix table statistics collecting error with metadata column predicate` does the title of the commit still stand?

I've reverted your changes and didn't encounter an error in Trino/ but rather an assertion error in `testCollectingStatisticsWithPathColumnPredicate`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 6 above)
Comment (chenjian2664) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:5544]: I don’t know all the reasons, but what I understand is that Avro is a row-based storage format and does not inherently support storing column-level statistics. Therefore, it's somewhat different from columnar storage formats like Parquet and ORC
Could you help to provide a concise comment

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 8]
```diff
@@ -2801,7 +2801,7 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
...
+                    return TableStatisticsReader.getTableStatistics(typeManager, session, handle, icebergTable, fileSystemFactory.create(session.getIdentity(), icebergTable.io().properties()));
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:2804]: this line became long. 
we put all arguments on one line, or each on separate line:

```java
return TableStatisticsReader.getTableStatistics(
        typeManager,
        session,
        handle,
        icebergTable,
        fileSystemFactory.create(session.getIdentity(), icebergTable.io().properties()));
```

**Review by findepi — APPROVED** [review]

**Comment by chenjian2664** [comment]
@findinpath @mosabua @ebyhr @findepi Thank you all for the help

**Pull request merged**